components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/JWTSignatureValidationUtils.java

```diff
@@ -0,0 +1,201 @@
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPublicKey;
+
+/**
+ * Utility methods for JWT Signature Validation
+ */
+public class JWTSignatureValidationUtils {
+
+    private static final Log log = LogFactory.getLog(JWTSignatureValidationUtils.class);
+
+    private static final String JWKS_URI = "jwksUri";
+    private static final String JWKS_VALIDATION_ENABLE_CONFIG = "JWTValidatorConfigs.Enable";
+    private static final String ENFORCE_CERTIFICATE_VALIDITY
+            = "JWTValidatorConfigs.EnforceCertificateExpiryTimeValidity";
+
+    /**
+     * Method to validate the signature of the JWT
+     *
+     * @param signedJWT signed JWT whose signature is to be verified
+     * @param idp       Identity provider who issued the signed JWT
+     * @return whether signature is valid, true if valid else false
+     * @throws JOSEException
+     * @throws IdentityOAuth2Exception
+     */
+    public static boolean validateSignature(SignedJWT signedJWT, IdentityProvider idp)
+            throws JOSEException, IdentityOAuth2Exception {
+
+        boolean isJWKSEnabled = false;
+        boolean hasJWKSUri = false;
+        String jwksUri = null;
+
+        String isJWKSEnalbedProperty = IdentityUtil.getProperty(JWKS_VALIDATION_ENABLE_CONFIG);
+        isJWKSEnabled = Boolean.parseBoolean(isJWKSEnalbedProperty);
+        if (isJWKSEnabled) {
+            if (log.isDebugEnabled()) {
+                log.debug("JWKS based JWT validation enabled.");
+            }
+        }
+
+        IdentityProviderProperty[] identityProviderProperties = idp.getIdpProperties();
+        if (!ArrayUtils.isEmpty(identityProviderProperties)) {
+            for (IdentityProviderProperty identityProviderProperty : identityProviderProperties) {
+                if (StringUtils.equals(identityProviderProperty.getName(), JWKS_URI)) {
+                    hasJWKSUri = true;
+                    jwksUri = identityProviderProperty.getValue();
+                    if (log.isDebugEnabled()) {
+                        log.debug("JWKS endpoint set for the identity provider : " + idp.getIdentityProviderName() +
+                                ", jwks_uri : " + jwksUri);
+                    }
+                    break;
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("JWKS endpoint not specified for the identity provider : " + idp
+                                .getIdentityProviderName());
+                    }
+                }
+            }
+        }
+
+        if (isJWKSEnabled && hasJWKSUri) {
+            JWKSBasedJWTValidator jwksBasedJWTValidator = new JWKSBasedJWTValidator();
+            return jwksBasedJWTValidator.validateSignature(signedJWT.getParsedString(), jwksUri, signedJWT.getHeader
+                    ().getAlgorithm().getName(), null);
+        } else {
+            JWSVerifier verifier = null;
+            JWSHeader header = signedJWT.getHeader();
```

Comment:
can we break this method into two parts? validateSignnatureWithJWKS and  validateSignatureWithCertificate ?

Reply:
fixed in e05e01f